Xaiki/shortcut

xaiki opened the pull request:
This branch implements removing the store and assuming that a file is downloaded if found on disk

xaiki committed:
base: introduce find_longest_prefix to factor code in

this now allows us to find super-prefixes for the last compoment as well as exact under-prefixes for the other

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
base: fixup, _cb_registery is index via cb.name

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: keep a reference to the fd we sendout for failure recovery

we'd leak fds if we don't do that

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
chunks: allow to resize segments on the fly

for some content, like the shards we get from amazon, we don't have reliable size to pre-process, for now we'll just resize segments until we get a complete signal

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus: slightly change API for request interest to return name and use it

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus: get name from requestInterest call response

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
file: rework flow and logic to allow for name changes

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
parallel: port changes in file, now we take a directory as a default entry point

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
http/manifest: little rework to use the new rename code

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
segments: split _open_file to make code moar readable

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
subscription: use file's directory entrypoint, and stop assuming we know the returned filename

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: have return functions return True for methods not to fail

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
better logging

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
subscription: fix passing consumer._filename to on_complete as it can change now

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
simple_store: port code to work with dbus producers

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
simple_store: tests

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
simple_store: fix name being a Name (convert to str)

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
base: TRY creating a new dbus connection for each DBusProducerSingleton

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
fixup: add correct flags

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
demo: don't remove old data to test re-entrance

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: simple _send_chunk implementation

this make simple_store work with dbus code path

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
demo: run with pure python (drop cprofile)

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
parallel: emit 'worker-complete' signal

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: fixup don't pass name twice

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
subscriptions: FIXUP: store _store_dir and _tmp_dir as realpath for substitution to work

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: basic (hacky) error handeling

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: try cached _pending_interest first

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: less logging

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: init workers where they belong

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: properly raise NotImplementedError()

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: log ETRYAGAIN

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
ostree_store: don't run store

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: pending_interest table is single source of truth

kill _wants_start

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus: simplify request_interest code path and centralize error handeling

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: encode try_again in pending_interest table and implement Timeout in _complete

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
demo: fix, implement runpython

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
file: implement shortcut when downloading and can find a file that looks right

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: fixup return_value doesn't take name twice it's auto folded

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
file: harder check for 'file looking right'

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

cosimoc commented on eos_data_distribution/ndn/file.py:
```diff
@@ -186,21 +186,71 @@ def _on_file_monitor_changed(self, monitor, file, other_file, event_type):
+
+    def setName(self, name):
+        logger.debug('set name: %s ↔ %s', name, self.name)
+        if str(name) == str(self.name):
```
This looks like it could be useful in the superclass too

xaiki replied:
not like this, because it reads the segment file that makes no sense for a generic File.Consumer (opposed to a File.FileConsumer)

cosimoc commented on eos_data_distribution/ndn/file.py:
```diff
@@ -186,21 +186,71 @@ def _on_file_monitor_changed(self, monitor, file, other_file, event_type):
         except ValueError as e:
             pass
 
+    def _read_segments_from_file(self, name):
```
You should not need to pass the name again here -- you can use `self.name`

xaiki replied:
fixd

cosimoc commented on eos_data_distribution/ndn/file.py:
```diff
@@ -186,21 +186,71 @@ def _on_file_monitor_changed(self, monitor, file, other_file, event_type):
             pass
 
+    def _read_segments_from_file(self, name):
+        if not self._filename:
```
Looks like this should go in the constructor instead

xaiki replied:
fixd

cosimoc commented on eos_data_distribution/ndn/http.py:
```diff
@@ -156,9 +156,10 @@ def _consume_queue(self):
-        self._getter = Getter(url, session=session,
+    def __init__(self, name, url, getter=None, session=None, *args, **kwargs):
+        getter = getter or Getter(url, session=session,
                               onData=lambda d: self.sendFinish(d))
```
Indentation a bit odd

xaiki replied:
fixd

cosimoc commented on eos_data_distribution/subscription.py:
```diff
@@ -95,7 +96,7 @@ def _fetch_manifest_complete(self, consumer):
         response = {
             "subscription_id": self.subscription_id,
-            "manifest_path": self._manifest_filename,
+            "manifest_path": self._manifest_consumer._filename,
```
Does not look like `self._manifest_consumer` exists at this point

cosimoc commented on eos_data_distribution/subscription.py:
```diff
@@ -62,14 +62,15 @@ def start(self):
     def _fetch_manifest_complete(self, consumer):
-        with open(self._manifest_filename, 'r') as f:
+        # we can't know the filename in advance, as it's timestamped
+        with open(consumer._filename, 'r') as f:
```
Would be nice to have a getter instead of accessing private consumer data

xaiki replied:
fixd

cosimoc commented on eos_data_distribution/ndn/dbus/base.py:
```diff
@@ -251,7 +251,9 @@ def __init__(self, name, dbus_name, skeleton):
 
-        self.con = Gio.bus_get_sync(BUS_TYPE, None)
+        address = Gio.dbus_address_get_for_bus_sync(BUS_TYPE, None)
+        self.con = Gio.DBusConnection.new_for_address_sync(address, Gio.DBusConnectionFlags.NONE, None, None)
```
This is really odd! Do things not work without it? I'd expect the address to return exactly the same value every time.

xaiki replied:
we used to need 2 DBus connections because if i reused the same one to publish on 2 different interfaces everything got exported to both.
this code is not strictly required now, but it's more correct than the former one.

cosimoc commented on eos_data_distribution/ndn/dbus/base.py:
```diff
@@ -251,9 +251,9 @@ def __init__(self, name, dbus_name, skeleton):
-        self.con = Gio.DBusConnection.new_for_address_sync(address, Gio.DBusConnectionFlags.NONE, None, None)
 
+        address = Gio.dbus_address_get_for_bus_sync(BUS_TYPE, None)
+        self.con = Gio.DBusConnection.new_for_address_sync(address, Gio.DBusConnectionFlags.AUTHENTICATION_CLIENT | Gio.DBusConnectionFlags.MESSAGE_BUS_CONNECTION, None, None)
```
Maybe squash with previous? (Even though it's still unclear why it's needed.)

xaiki replied:
from the docs: 
```
The returned object is a singleton, that is, shared with other callers of g_bus_get() and g_bus_get_sync() for bus_type . In the event that you need a private message bus connection, use g_dbus_address_get_for_bus_sync() and g_dbus_connection_new_for_address().
```

in store we have the base connection and we need another connection for the chunks code so that they are properly separated

xaiki replied:
as for squashing... a lot of things must be squashed in these commits =/

cosimoc commented on eos_data_distribution/ndn/dbus/base.py:
```diff
@@ -199,7 +199,13 @@ def expressInterest(self, interest=None, try_again=False):
         if not interest:
             interest = self.name.toString()
 
-        self.interest = interest
```
Is there nothing that relies on `self.interest` to be set?

xaiki replied:
no it's an old API, we now support many interests (for re-entrance)

cosimoc commented on eos_data_distribution/ndn/dbus/base.py:
```diff
@@ -184,9 +184,9 @@ def _on_manager_ready(self, proxy, res):
-        for interest, dbus_path, dbus_name in self._pending_interests.values():
-             if self._dbus_express_interest(interest, dbus_path, dbus_name):
-                del self._pending_interests[interest]
+        for interest in self._pending_interests.keys():
```
You could get the keys beforehand, since you use it twice in this function (or move the debug inside the loop)

xaiki replied:
fixd

cosimoc commented on eos_data_distribution/ndn/dbus/base.py:
```diff
@@ -244,9 +244,13 @@ def _on_request_interest_complete(self, interface, res, interest):
             logger.debug('request-interest got error: %s', error)
             if str(error).find('ETRYAGAIN') != -1:
                 return self._dbus_express_interest(interest)
+            elif error.code == 24: # Timeout
```
Can you use `error.matches()` instead? That would allow disambiguating the error domain as well (the code alone is not enough to uniquely identify the error)

xaiki replied:
i'm very much not sure how to use that API...
```
In [19]: int(Gio.DBusError.SPAWN_FORK_FAILED)
Out[19]: 24
```
that said when the error gets in i get:
```
GLib.Error: g-io-error-quark: Timeout was reached (24)
```

cosimoc reviewed (changes requested):
Thanks, @xaiki. Left some comments here.

cosimoc commented:
(Make sure to also look at the comments GitHub thinks are outdated, because not all of them may be.)

xaiki committed:
http: fix identation

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: rework consumer init to be robust

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: exit early if no final_segment in _on_progress

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
file: add filename getter to FileProducer

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/base: don't process .keys() twice in flush_pending_interests

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
file: have _read_segments_from_file use self.name so it doesn't require an argument

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki commented on eos_data_distribution/ndn/file.py:
```diff
@@ -186,21 +186,71 @@ def _on_file_monitor_changed(self, monitor, file, other_file, event_type):
             pass
 
+    def _read_segments_from_file(self, name):
+        if not self._filename:
```
how so ? the trick is that we re-use it when ever the name changes this is why it was split out

xaiki committed:
file: split setting filename to constructor

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
implement DiscoverVersion

this requires killing the filename argument in file as we *need* to get it from the DiscoverVersion call
it also changes the way chunks works and introduce a new _complete call to abort early

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus: rename request_interest to express_interest

this makes the chunks codepath clearer as express_interest first goes out to resolve_version and then handles the request_interest call.
error handeling is hence done in base.py and on express_interest

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

cosimoc commented on eos_data_distribution/ndn/dbus/chunks.py:
```diff
@@ -201,15 +222,17 @@ def _on_progress(self, proxy, name, first_segment, last_segment):
 
     def _request_interest_complete(self, interface, res, interest):
         logger.info('Consumer: request interest complete: %s, %s, %s', interface, res, interest)
         name, final_segment, fd_list = interface.call_request_interest_finish(res)
```
Since you have removed the calls to `setName()` and `_set_final_segment()` here and we don't do anything with the return values, should something be changed in the dbus interface itself? It strikes me as odd.

xaiki replied:
it's a bit strange indeed, but it's left to have the same API as Base

xaiki commented:
(waiting for merge as we talked on today's 1:1)

xaiki committed:
file: add sha256 checksum

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: remove unused self.interest in consumer

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
subscription: use csum

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: emit interest in ResolveVersion so that subscription keeps working

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
dbus/chunks: fix typo

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>

xaiki committed:
pit implementation

Signed-off-by: Niv Sardi <xaiki@evilgiggle.com>
